### PR TITLE
backendutil: Merge Match* functions

### LINKED ISSUE #222: Unexpected result of MatchFlags for SearchCriteria with Not
Here is minimal test case:
```
import (
	"fmt"

	"github.com/emersion/go-imap"
	"github.com/emersion/go-imap/backend/backendutil"
)

func main() {
	crit := imap.SearchCriteria{
		Not: []*imap.SearchCriteria{
			&imap.SearchCriteria{
				Body: []string{"hi!"},
			},
		},
	}

	fmt.Println(backendutil.MatchFlags([]string{"FlagA", "FlagB"}, &crit))
}
```

Since there is no With(out)Flags I expect MatchFlags to return true but it returns false instead:
```
⟩ go run test.go
false
```

### PR DESCRIPTION
**Pull request by foxcpp**
Backward-compatibility-breaking change.
Closes #222.

### CODE REVIEW

File: backend/backendutil/search_test.go [hunk 1]
```diff
@@ -10,8 +10,14 @@ import (
...
 )
 
+var testInternalDate = time.Unix(1483997966, 0)
+
 var matchTests = []struct {
 	criteria *imap.SearchCriteria
+	seqNum   uint32
+	uid      uint32
+	date     time.Time
+	flags    []string
```
Comment (emersion) [inline backend/backendutil/search_test.go]: Style: alignment issue here

Comment (foxcpp) [reply backend/backendutil/search_test.go]: Ugh, gofmt seems to be fine with it. How it should be formatted?

Comment (emersion) [reply backend/backendutil/search_test.go]: Oh, if gofmt is fine with that, let's keep it that way

Comment (emersion) [reply backend/backendutil/search_test.go]: Argh, I see: for some reason GitHub displays `fl` as a ligature in one character. Thus `[]string` is misaligned. That's pretty weird.

**Comment by emersion** [comment]
Thanks!

**Pull request merged**